src/CodeGen_Xtensa.cpp

```diff
@@ -633,16 +633,13 @@ HALIDE_ALWAYS_INLINE void store_variable<native_vector_u8, uint8_t, VECTOR_WIDTH
     IVP_SAPOS2NX8U_FP(align, ptr);
 }
 
-template <typename VectorType, typename OffsetType, typename BaseType, int Lanes>
-HALIDE_ALWAYS_INLINE VectorType gather_load(const void *base, const OffsetType& offset) {
-    BaseType __attribute__((aligned(XCHAL_VISION_SIMD8))) tmp[Lanes];
-    int offsets[Lanes];
-    store<OffsetType, int32_t, Lanes>(offset, &offsets[0], 0);
-    for (int i = 0; i < Lanes; i++) {
-        tmp[i] = ((const BaseType*)base)[offsets[i]];
-    }
-
-    return *((VectorType *)tmp);
+template <typename VectorType, typename OffsetType, typename BaseType, int Lanes, bool isTCM>
+HALIDE_ALWAYS_INLINE VectorType gather_load(const void *base, const OffsetType& offset){
+  if (isTCM) {
```

Comment:
We probably don't want to have an if statement here. I think slight modification of your approach might fix it: I would keep a single gather_load function (instead of gather_load_tcm and gather_load_default), add a new template parameter isTCM (already there), all template specializations which are implemented with intrinsics will have it set to true and then emit additional true or false template parameter at the handler of Allocate node. This way only gather_loads with matching types AND is_tcm parameter set to true will get their corresponding specialization called, but everything else will default to scalarized implementation.

Reply:
One comment, here. Since we emit gather_load during visiting Load node, I think we should emit is_tcm aslo there and not in Allocate node.